keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -335,34 +335,47 @@ The versions of CNI plugin binaries that must be used for proper dual-stack func
+1. If the user choose to set `.spec.ipFamily` then `.status.preferredIPFamily` value will be one entry equal to the value of `.spec.ipFamily`.
+2. If the value of `.spec.ipFamily` is unset then the value of `.status.preferredIPFamily` will be set according to the value set in `--services-cluster-ip-range` flag. This means during a cluster upgrade (from non-dualstack to dualstack) the value of this feild will be automatically set to `--services-cluster-ip-range`. 
 
-For example, the following Service specification includes the `spec.ipFamily` field. Kubernetes will assign an IPv6 address (also known as a “cluster IP”) from the configured service-cluster-ip-range to this Service.
+
+For example, Kubernetes assigned an IPv4 address for `.spec.clusterIP` and assigned `['IPv4' , 'IPv6']` for `.status.preferredIPFamily` because this service was created with `.spec.ipFamily=nil` and the cluster was configured with `--services-cluster-ip-range=<IPv4 CIDR>,<IPv6 CIDR>`
 
 ```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
-  ipFamily: IPv6
+  type: ClusterIP
+  clusterIP: 1.2.3.4
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+status:
+  preferredIPFamily:
+    - IPv4
+    - IPv6
 ```
 
+> A dualstack cluster can be configured `--services-cluster-ip-range=<single service cidr>` flag in this case the value of `.status.ipFamily` will always be a single entry pointing to the address of the cidr configured in `--services-cluster-ip-range` flag. Attempting to select an ip family in `.spec.ipFamily` that is not configured on the cluster will generate a validation error on the cluster.
+
 Dual-stack services (having both an IPv4 and IPv6 cluster IP) are currently out of scope of this proposal. Should there be use-cases where dual-stack Services are needed then we can revisit.
 
 #### Type Headless services
+Headless services will operate using the following conditions:
+1. api-server does not default `ipFamily` field. User may optionally to choose to set it.
+2. If `ipFamily` field is set by the user, then the value of `preferredIPFamily` will follow accordingly. Example: setting `ipFamily` to `IPv6` then `preferredIPFamily` will have `IPv6` as a value.
+3. If `ipFamily` field is not set, then the value of `preferredIPFamily` will carry the values set in `--service-cluster-ip-range`, which can be a single value in even in a dualstack cluster.
 
-\<TBD\>
 
 #### Type ExternalName
 
-\<TBD\>
+Since ips and their families are not involved in services with type `ExternalName` the values of `.spec.ipFamily` and `.status.preferredIPFamily` is expected to be `nil`. Attempting to set the values to anything else will not be accepted by the api-server.
```

Comment:
Compat problem:  Today a user can RMW Service object to change `type` from "ClusterIP" -> "ExternalName" and it works.  If we do not auto-clear `ipFamily` in that case, this operation will start failing.  I think we have to just clear those fields.